AutoCollection/WebSnippet.ts

```diff
@@ -0,0 +1,146 @@
+import * as path from "path";
+import http = require("http");
+import fs = require('fs');
+
+import Logging = require("../Library/Logging");
+import TelemetryClient = require("../Library/TelemetryClient");
+
+class WebSnippet {
+
+    public static INSTANCE: WebSnippet;
+
+    private static _snippet: string;
+    private static _aiUrl: string;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+
+    constructor(client: TelemetryClient) {
+        if (!!WebSnippet.INSTANCE) {
+            throw new Error("Web snippet injection should be configured from the applicationInsights object");
+        }
+
+        WebSnippet.INSTANCE = this;
+        // AI URL used to validate if snippet already included
+        WebSnippet._aiUrl = " https://js.monitor.azure.com/scripts/b/ai.2";
+
+        let snippetPath = path.resolve(__dirname, "../../node_modules/applicationinsight-web-snippet/snippet/snippet.min.js");
+        if (client.config.isDebugWebSnippet) {
+            snippetPath = path.resolve(__dirname, "../../node_modules/applicationinsight-web-snippet/snippet/snippet.js");
+        }
+        fs.readFile(snippetPath, function (err, snippet) {
+            if (err) {
+                Logging.warn("Failed to load AI Web snippet. Ex:" + err);
+            }
+            WebSnippet._snippet = snippet.toString().replace("INSTRUMENTATION_KEY", client.config.instrumentationKey);
+        });
+    }
+
+    public enable(isEnabled: boolean) {
+        this._isEnabled = isEnabled;
+
+        if (this._isEnabled && !this._isInitialized) {
+            this._initialize();
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    private _initialize() {
+        this._isInitialized = true;
```

Comment:
Ideally, you would set this to true after all the initialization work is down.  There is a bunch of code and if anything fails and we bail halfway through you don't want isInitialized lying!

Reply:
follow the previous original pattern [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/033d1f4871ba9a5c2b136f98cebc7a21dd91ff37/AutoCollection/Exceptions.ts#L39)

Reply:
Would it hurt to move this to end of the method?  If not I suggest you move it :)

Reply:
I assume putting `isInitialized` flag in the beginning would prevent double monkey patching. for example, when user enables the feature but initialization hasn't been finished and then user enabled it again, if isInitialized` flag is still be false at this time, the process will start a new monkey patch again. I am not sure if there is a better way to handle this scenario.😊